[BE][Attention] Code de-dup

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139791
* #139763
* __->__ #139784
* #139788

The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implementations are trivial oneline call
Also, as suggested by @Skylion007, replace `at::where(foo->logical_not, -inf, 0)` with `at::where(*foo, 0, -inf)`

malfet committed:
[BE][Attention] Code de-dup

The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implienetations are trivial oneline call

[ghstack-poisoned]

Skylion007 reviewed (commented):
optimization nit

malfet committed:
Update on "[BE][Attention] Code de-dup"


The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implienetations are trivial oneline call

[ghstack-poisoned]

malfet committed:
Update on "[BE][Attention] Code de-dup"


The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implienetations are trivial oneline call

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -524,35 +524,29 @@ inline void validate_sdpa_input(
   // to mask *out*.
   if (attn_mask->dtype() == at::kBool) {
-    return at::where(attn_mask->logical_not(), -std::numeric_limits<double>::infinity(), at::scalar_tensor(0.0, at::TensorOptions().dtype(dtype).device(attn_mask->device())));
+    return at::where(*attn_mask, 0.0, at::scalar_tensor(neg_inf, at::TensorOptions().dtype(dtype).device(attn_mask->device())));
```
we should probably also have 0.0 be a scalar tensor right? I am not sure why this didnt also need to be updated

Skylion007 replied:
Lol, I made this edit on the other PR in the stack.

malfet replied:
I suspect only one scalar_tensor is needed to preserve out dtype, other can be inferred from the first.

malfet committed:
Update on "[BE][Attention] Code de-dup"


The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implienetations are trivial oneline call

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet committed:
Update on "[BE][Attention] Code de-dup"


The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implementations are trivial oneline call
Also, as suggested by Skylion007, replace `at::where(foo->logical_not, -inf, 0)` with `at::where(*foo, 0, -inf)`

[ghstack-poisoned]

malfet commented:
@pytorchbot merge
